src/TurbulenceConvection/EDMF_functions.jl

```diff
@@ -0,0 +1,1352 @@
+        q_surf = q_surface_bc(surf, grid, state, edmf, i)
+        a_surf = area_surface_bc(surf, edmf, i)
+        prog_up[i].ρarea[kc_surf] = ρ_c[kc_surf] * a_surf
+        prog_up[i].ρaθ_liq_ice[kc_surf] = prog_up[i].ρarea[kc_surf] * θ_surf
+        prog_up[i].ρaq_tot[kc_surf] = prog_up[i].ρarea[kc_surf] * q_surf
+        if edmf.moisture_model isa NonEquilibriumMoisture
+            q_liq_surf = FT(0)
+            q_ice_surf = FT(0)
+            prog_up[i].ρaq_liq[kc_surf] = prog_up[i].ρarea[kc_surf] * q_liq_surf
+            prog_up[i].ρaq_ice[kc_surf] = prog_up[i].ρarea[kc_surf] * q_ice_surf
+        end
+        prog_up_f[i].ρaw[kf_surf] = ρ_f[kf_surf] * w_surface_bc(surf)
+        ae_surf -= a_surf
+    end
+
+    flux1 = surf.shf / cp
+    flux2 = surf.ρq_tot_flux
+    zLL::FT = grid.zc[kc_surf].z
+    ustar = surf.ustar
+    oblength = surf.obukhov_length
+    ρLL = prog_gm.ρ[kc_surf]
+    ρ_ae = ρ_c[kc_surf] * ae_surf
+    mix_len_params = mixing_length_params(edmf)
+    prog_en.ρatke[kc_surf] =
+        ρ_ae *
+        get_surface_tke(mix_len_params, surf.ustar, zLL, surf.obukhov_length)
+    if edmf.thermo_covariance_model isa PrognosticThermoCovariances
+        prog_en.ρaHvar[kc_surf] =
+            ρ_ae *
+            get_surface_variance(flux1 / ρLL, flux1 / ρLL, ustar, zLL, oblength)
+        prog_en.ρaQTvar[kc_surf] =
+            ρ_ae *
+            get_surface_variance(flux2 / ρLL, flux2 / ρLL, ustar, zLL, oblength)
+        prog_en.ρaHQTcov[kc_surf] =
+            ρ_ae *
+            get_surface_variance(flux1 / ρLL, flux2 / ρLL, ustar, zLL, oblength)
+    end
+    return nothing
+end
+
+function surface_helper(surf::SurfaceBase, grid::Grid, state::State)
+    FT = float_type(state)
+    kc_surf = kc_surface(grid)
+    prog_gm = center_prog_grid_mean(state)
+    zLL::FT = grid.zc[kc_surf].z
+    ustar = surf.ustar
+    oblength = surf.obukhov_length
+    ρLL = prog_gm.ρ[kc_surf]
+    return (; ustar, zLL, oblength, ρLL)
+end
+
+function a_up_boundary_conditions(surf::SurfaceBase, edmf::EDMFModel, i::Int)
+    a_surf = area_surface_bc(surf, edmf, i)
+    return (; bottom = CCO.SetValue(a_surf), top = CCO.Extrapolate())
+end
+
+function a_bulk_boundary_conditions(surf::SurfaceBase, edmf::EDMFModel)
+    N_up = n_updrafts(edmf)
+    a_surf = sum(i -> area_surface_bc(surf, edmf, i), 1:N_up)
+    return (; bottom = CCO.SetValue(a_surf), top = CCO.Extrapolate())
+end
+
+function a_en_boundary_conditions(surf::SurfaceBase, edmf::EDMFModel)
+    N_up = n_updrafts(edmf)
+    a_surf = 1 - sum(i -> area_surface_bc(surf, edmf, i), 1:N_up)
+    return (; bottom = CCO.SetValue(a_surf), top = CCO.Extrapolate())
+end
+
+function area_surface_bc(
+    surf::SurfaceBase{FT},
+    edmf::EDMFModel,
+    i::Int,
+)::FT where {FT}
+    N_up = n_updrafts(edmf)
+    return surf.bflux > 0 ? edmf.surface_area / N_up : FT(0)
+end
+
+function w_surface_bc(::SurfaceBase{FT})::FT where {FT}
```

Comment:
Need to revisit this: If we are solving for covariant vertical velocity, the BC is not zero at topography. And generally, we should be careful to distinguish between physical vertical velocity (w) and covariant vertical velocity (u_3) now.